### PR TITLE
[FLINK-13237][table-planner-blink] Add expression table api test to blink

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

Add expression table api test and fix bugs.

## Verifying this change

ut

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by wuchong** [comment]
Hi @JingsongLi , can you rebase this pull request?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala [hunk 1]
```diff
@@ -1198,583 +1498,992 @@ class ScalarFunctionsTest extends ScalarTypesTestBase {
...
+    testAllApis(
+      'f29.round('f30),
+      "f29.round(f30)",
       "ROUND(f29, f30)",
       0.5.toString)
 
-    testSqlApi(
-      "ROUND(f31, f7)",
-      "-0.123")
+    // Now not support round decimal with non-literal
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala]: Is it a bug in blink or we don't support it yet? Do we need to fix this before 1.9?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala [hunk 2]
```diff
@@ -2531,27 +3423,60 @@ class ScalarFunctionsTest extends ScalarTypesTestBase {
...
+      "2016-06-15".toDate + 1.day,
+      "'2016-06-15'.toDate + 1.day",
+      "TIMESTAMPADD(DAY, 1, DATE '2016-06-15')",
+      "2016-06-16")
 
+    testAllApis(
+      nullOf(DataTypes.TIMESTAMP(3)) + 3.months,
+      "Null(SQL_TIMESTAMP) + 3.months",
+      "TIMESTAMPADD(MONTH, 3, CAST(NULL AS TIMESTAMP))",
+      "null")
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala]: It seems that we missed some tests under this?

This is from flink planner expression test:
```scala
testAllApis(
      "2016-02-24 12:42:25".toTimestamp + nullOf(Types.INTERVAL_MILLIS),
      "'2016-02-24 12:42:25'.toTimestamp + nullOf(INTERVAL_MILLIS)",
      "TIMESTAMPADD(HOUR, CAST(NULL AS INTEGER), TIMESTAMP '2016-02-24 12:42:25')",
      "null")

    testAllApis(
      nullOf(Types.SQL_TIMESTAMP) + -200.hours,
      "nullOf(SQL_TIMESTAMP) + -200.hours",
      "TIMESTAMPADD(HOUR, -200, CAST(NULL AS TIMESTAMP))",
      "null")

    testAllApis(
      nullOf(Types.SQL_TIMESTAMP) + 3.months,
      "nullOf(SQL_TIMESTAMP) + 3.months",
      "TIMESTAMPADD(MONTH, 3, CAST(NULL AS TIMESTAMP))",
      "null")

    // TIMESTAMPADD with DATE returns a TIMESTAMP value for sub-day intervals.
    testAllApis("2016-06-15".toDate + 1.month,
      "'2016-06-15'.toDate + 1.month",
      "timestampadd(MONTH, 1, date '2016-06-15')",
      "2016-07-15")

    testAllApis("2016-06-15".toDate + 1.day,
      "'2016-06-15'.toDate + 1.day",
      "timestampadd(DAY, 1, date '2016-06-15')",
      "2016-06-16")

    testAllApis("2016-06-15".toTimestamp - 1.hour,
      "'2016-06-15'.toTimestamp - 1.hour",
      "timestampadd(HOUR, -1, date '2016-06-15')",
      "2016-06-14 23:00:00.0")

    testAllApis("2016-06-15".toTimestamp + 1.minute,
      "'2016-06-15'.toTimestamp + 1.minute",
      "timestampadd(MINUTE, 1, date '2016-06-15')",
      "2016-06-15 00:01:00.0")

    testAllApis("2016-06-15".toTimestamp - 1.second,
      "'2016-06-15'.toTimestamp - 1.second",
      "timestampadd(SQL_TSI_SECOND, -1, date '2016-06-15')",
      "2016-06-14 23:59:59.0")

    testAllApis("2016-06-15".toTimestamp + 1.second,
      "'2016-06-15'.toTimestamp + 1.second",
      "timestampadd(SECOND, 1, date '2016-06-15')",
      "2016-06-15 00:00:01.0")

    testAllApis(nullOf(Types.SQL_TIMESTAMP) + 1.second,
      "nullOf(SQL_TIMESTAMP) + 1.second",
      "timestampadd(SECOND, 1, cast(null as date))",
      "null")

    testAllApis(nullOf(Types.SQL_TIMESTAMP) + 1.day,
      "nullOf(SQL_TIMESTAMP) + 1.day",
      "timestampadd(DAY, 1, cast(null as date))",
      "null")

    // Round to the last day of previous month
    testAllApis("2016-05-31".toDate + 1.month,
      "'2016-05-31'.toDate + 1.month",
      "timestampadd(MONTH, 1, date '2016-05-31')",
      "2016-06-30")

    testAllApis("2016-01-31".toDate + 5.month,
      "'2016-01-31'.toDate + 5.month",
      "timestampadd(MONTH, 5, date '2016-01-31')",
      "2016-06-30")

    testAllApis("2016-03-31".toDate - 1.month,
      "'2016-03-31'.toDate - 1.month",
      "timestampadd(MONTH, -1, date '2016-03-31')",
      "2016-02-29")

    testAllApis("2016-03-31".toDate - 1.week,
      "'2016-03-31'.toDate - 1.week",
      "timestampadd(WEEK, -1, date '2016-03-31')",
      "2016-03-24")
```

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala [hunk 3]
```diff
@@ -2785,7 +3766,9 @@ class ScalarFunctionsTest extends ScalarTypesTestBase {
 
   @Test
   def testNullBigDecimal(): Unit = {
-    testSqlApi(
+    testAllApis(
+      ExpressionParser.parseExpression("f41.sign()"),
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala]: `'f41.sign()` ?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala [hunk 4]
```diff
@@ -2785,7 +3766,9 @@ class ScalarFunctionsTest extends ScalarTypesTestBase {
...
+      "f41.sign()",
       "SIGN(f41)",
       "null")
     testSqlApi("SIGN(f41)", "null")
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala]: Remove this?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/utils/ExpressionTestBase.scala [hunk 5]
```diff
@@ -162,31 +169,34 @@ abstract class ExpressionTestBase {
           }
 
           assertEquals(
-            s"Wrong result for: [$originalExpr] optimized to: [$optimizedExpr]",
+            s"Wrong result for: [$optimizedExpr]",
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/utils/ExpressionTestBase.scala]: table api will be null.
but I can do some if else to keep it.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/NonDeterministicTests.scala [hunk 6]
```diff
@@ -27,47 +28,64 @@ import org.junit.{Ignore, Test}
 /**
   * Tests that can only be checked manually as they are non-deterministic.
   */
-@Ignore
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/NonDeterministicTests.scala]: Just keep same to flink, More convenient comparison.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/utils/ExpressionTestBase.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/utils/ExpressionTestBase.scala]: Why table api will be null?  Couldn't we use `Expression.toString()` ?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala (hunk 1 above)
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala]: see `ReturnTypeInference.inferRound`, now we have copied calcite's type infer for round. It require round.right must be a literal...

Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala]: I don't think this is a bug

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala (hunk 2 above)
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala]: Thanks~

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/utils/ExpressionTestBase.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/utils/ExpressionTestBase.scala]: Good, Expression.asSummaryString is OK.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala (hunk 2 above)
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/ScalarFunctionsTest.scala]: There are some limitation in blink... I think we need fix it. I will add TODO and track it in jira

**Comment by wuchong** [comment]
Thanks @JingsongLi , this looks good to me now. 

However, there are cases failed on travis, could you have a look again?

**Comment by JingsongLi** [comment]
@wuchong I ignored this cases and created JIRA: https://issues.apache.org/jira/browse/FLINK-13580
I think it can be supported in 1.10.

**Comment by wuchong** [comment]
Thanks @JingsongLi .

**Comment by wuchong** [comment]
Merging...

**Pull request closed without merging**